### PR TITLE
Baseline/Quest

### PR DESCRIPTION
**Pull request by sahiljoshi515**
# 🧱 Baseline Addition PR

## 📘 Description

Baseline: Quest
Paper:  [Quest: Query-Aware Sparsity for Efficient Long-Context LLM Inference
](https://arxiv.org/abs/2408.07092)

---

## ✅ Baseline Checklist 
(Follow the docs/baslines/developer_guide.md)

| Item | Status | Notes (Note comments, remove placehoders) |
|------|---------|-------|
| 🧩 **1. baseline code** | [ N]  |  The code was written by understanding paper's low-level idea. |
| 🧪 **2. Correctness**  | | One of the two needs to be done ( a or b) |
| &nbsp;&nbsp;&nbsp; a. Code matches reference implementation | [Y ] |  test_quest_sparsity_correctness stress tests code match|
| &nbsp;&nbsp;&nbsp; b. Results reproduced by matching evaluation setup | [N ] |  |
| 🧭 **3. Explanation of Differences** | [ Y] |The evaluation setup is different in authors code and ours (1. dense layers in first 2 blocks) |
| ⚙️ **4. Performance Profiled** | [ N] |  153.1% overhead |
| 📖 **5. Documentation Created** | [Y ] | `docs/baselines/Quest.md` |

---

### quality
Code match.
```
pytest tests/unit/sparse_attention/research_attention/maskers/fixed/implementations/test_quest_sparsity_correctness.py

============================================================== 36 passed, 1 warning in 108.82s (0:01:48) ==============================================================
```

results in different setups

We use multifieldqa, hotpotqa, gov_report and qasper from Longbench to compare our and author's results since author's code supports Longbench. 

Our Results:

| Dataset | Token Budget 256 | Token Budget 512 | Token Budget 1024 | Token Budget 2048 | Token Budget 4096 |
|:--------|:---------------:|:---------------:|:---------------:|:---------------:|:---------------:|
| **Qasper** | 17.11 | 24.4 | 29.96 | 32.65 | 30.51 |
| **HotpotQA** | 13.7 | 27.07 | 28.72 | 37.11 | 38.55 |
| **MultifieldQA** | 26.49 | 37.52 | 44 | 47.52 | 44.23 |
| **GovReport** | 10.86 | 16.58 | 24.02 | 26.73 | 30.14 |

Their Results:

Screenshot 2025-10-29 at 09 32 40

### research performance   (A100)

🔍 Comparative Analysis:
============================================================
📈 Masker Overhead Analysis:

    Sparse Attention: 1.738 ms
    Baseline (no maskers): 0.686 ms
    Masker overhead: 1.051 ms (153.1%)
    ❌ High overhead - significant masker cost

============================================================
```

## 🧩 Additional Notes
NA

### CODE REVIEW

File: sparse_attention_hub/sparse_attention/research_attention/maskers/fixed/implementations/quest_top_k.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+        if pad_k > 0:
+            ap = F.pad(allowed_prob, (0, pad_k), value=0.0)
+        else:
+            ap = allowed_prob
+
+        ap = ap.view(B, 1, Q, num_pages, page_size)  # [B,1,Q,P,ps]
+        return (ap.max(dim=-1).values > 0).squeeze(1)  # [B,Q,P] bool
+
+    def _pages_to_token_mask(self, topk_pages: torch.Tensor, K: int, page_size: int, device: torch.device, dtype: torch.dtype) -> torch.Tensor:
```
Comment (apd10) [inline sparse_attention_hub/sparse_attention/research_attention/maskers/fixed/implementations/quest_top_k.py]: can move to utility

File: sparse_attention_hub/sparse_attention/research_attention/maskers/fixed/implementations/quest_top_k.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        am = attention_mask[..., :K]
+        if am.dtype == torch.bool:
+            allowed = (am == 0).to(torch.float32)
+        else:
+            allowed = (am >= 0).to(torch.float32)
+        if allowed.dim() == 3:
+            allowed = allowed.unsqueeze(1)  # [B,1,*,K]
+        return allowed
+
+    def _page_any_valid(self, allowed_prob: torch.Tensor, page_size: int, num_pages: int) -> torch.Tensor:
```
Comment (apd10) [inline sparse_attention_hub/sparse_attention/research_attention/maskers/fixed/implementations/quest_top_k.py]: can move to utility

**Review by apd10 — CHANGES_REQUESTED** [review]

**Review by apd10 — APPROVED** [review]
